gwpy/io/kerberos.py

```diff
@@ -142,42 +167,111 @@ def kinit(
                             "a ticket, or consider using a keytab file")
 
     # get credentials
-    if realm is None:
-        realm = 'LIGO.ORG'
     if username is None:
         verbose = True
         username = input(
-            f"Please provide username for the {realm} kerberos realm: ",
+            "Kerberos principal (user@REALM): ",
         )
-    identity = f"{username}@{realm}"
+    if "@" not in username:
+        username = f"{username}@{DEFAULT_REALM}"
+    principal = gssapi.Name(
+        base=username,
+        name_type=gssapi.NameType.kerberos_principal,
+    )
     if not keytab and password is None:
         verbose = True
-        password = getpass.getpass(prompt=f"Password for {identity}: ")
+        password = getpass.getpass(prompt=f"Password for {principal}: ")
 
-    # format kinit command
-    if keytab:
-        cmd = [exe, '-k', '-t', keytab, identity]
-    else:
-        cmd = [exe, identity]
-    if krb5ccname:
-        krbenv = {'KRB5CCNAME': krb5ccname}
-    else:
-        krbenv = None
-
-    # execute command
-    kget = subprocess.Popen(cmd, env=krbenv, stdout=subprocess.PIPE,
-                            stdin=subprocess.PIPE)
-    if not keytab:
-        kget.communicate(password.encode('utf-8'))
-    kget.wait()
-    retcode = kget.poll()
-    if retcode:
-        raise subprocess.CalledProcessError(kget.returncode, ' '.join(cmd))
+    # generate credential
+    acquire_kw = {  # common options for acquire methods
+        "ccache": ccache,
+        "lifetime": lifetime,
+    }
+    try:
+        if keytab:
+            creds = _acquire_keytab(principal, str(keytab), **acquire_kw)
+        else:
+            print(principal, password, acquire_kw)
+            creds = _acquire_password(principal, password, **acquire_kw)
+    except gssapi.exceptions.GSSError:
+        raise KerberosError(
+            f"failed to generate Kerberos TGT for {principal}",
+        )
     if verbose:
-        print(f"Kerberos ticket generated for {identity}")
+        print(
+            f"Kerberos ticket acquired for {creds.name} "
+            f"({creds.lifetime} seconds remaining)",
+        )
+
+
+def _acquire_keytab(principal, keytab, ccache=None, lifetime=None):
+    """Acquire a Kerberos TGT using a keytab.
+    """
+    import gssapi
```

Comment:
Yes, because we need the `gssapi` module as a local reference in this function. The `try/except` above ensures that the `import` will work but doesn't make the `gssapi` object available to children.